# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [62]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
import datetime
import seaborn as sns

path="/Users/atousachaharsoughshirazi/desktop/Bootcamp/Unit_6/data"
os.chdir(path) 
os.getcwd()

'/Users/atousachaharsoughshirazi/Desktop/Bootcamp/Unit_6/data'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [63]:
df=pd.read_csv('step3_output.csv')
df.drop(df.columns[0], axis=1, inplace=True)
#df.head()
df.state.value_counts()

New York          33
Michigan          29
Colorado          22
California        21
Pennsylvania      19
New Hampshire     16
Wisconsin         16
Vermont           15
Minnesota         14
Utah              13
Idaho             12
Montana           12
Massachusetts     11
Oregon            10
Washington        10
Maine              9
New Mexico         9
Wyoming            8
North Carolina     6
Ohio               5
Connecticut        5
Nevada             4
West Virginia      4
Illinois           4
Virginia           4
Iowa               3
Alaska             3
Indiana            2
Arizona            2
New Jersey         2
South Dakota       2
Missouri           2
Maryland           1
Tennessee          1
Rhode Island       1
Name: state, dtype: int64

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [64]:
 # select object type columns
dfo=df['state'].values
df = pd.concat([df.drop(columns='state', axis=1), pd.get_dummies(dfo, drop_first=True)], axis=1)
#df = pd.concat([df.drop([df.state], axis=1), pd.get_dummies(df.state)], axis=1)
df.head(-5)

Name  summit_elev  vertical_drop  base_elev  trams  \
0          Alyeska Resort         3939           2500        250      1   
1     Eaglecrest Ski Area         2600           1540       1200      0   
2        Hilltop Ski Area         2090            294       1796      0   
3        Arizona Snowbowl        11500           2300       9200      0   
4     Sunrise Park Resort        11100           1800       9200      0   
..                    ...          ...            ...        ...    ...   
320     Whitecap Mountain         1750            400       1295      0   
321       Wilmot Mountain         1030            230        800      0   
322  Grand Targhee Resort         9920           2270       7851      0   
323         Hogadon Basin         8000            640       7400      0   
324          Jackson Hole        10450           4139       6311      3   

     fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0          0.0          0          2     2       0  ...             0   
1          0.0          0          0     0       0  ...             0   
2          0.0          0          0     0       1  ...             0   
3          0.0          1          0     2       2  ...             0   
4          0.0          0          1     2       3  ...             0   
..         ...        ...        ...   ...     ...  ...           ...   
320        0.0          0          0     1       0  ...             0   
321        0.0          0          0     3       2  ...             0   
322        0.0          0          2     2       0  ...             0   
323        0.0          0          0     0       0  ...             0   
324        0.0          0          4     4       1  ...             0   

     South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0               0          0     0        0         0           0   
1               0          0     0        0         0           0   
2               0          0     0        0         0           0   
3               0          0     0        0         0           0   
4               0          0     0        0         0           0   
..            ...        ...   ...      ...       ...         ...   
320             0          0     0        0         0           0   
321             0          0     0        0         0           0   
322             0          0     0        0         0           0   
323             0          0     0        0         0           0   
324             0          0     0        0         0           0   

     West Virginia  Wisconsin  Wyoming  
0                0          0        0  
1                0          0        0  
2                0          0        0  
3                0          0        0  
4                0          0        0  
..             ...        ...      ...  
320              0          1        0  
321              0          1        0  
322              0          0        1  
323              0          0        1  
324              0          0        1  

[325 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [65]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [66]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
X_train

array([[-0.15866985, -0.91441494,  0.04048329, ..., -0.11076976,
        -0.22573306, -0.15762208],
       [-0.25518791,  1.47351277, -0.75311775, ..., -0.11076976,
        -0.22573306, -0.15762208],
       [-0.97639234, -0.9672452 , -0.87521021, ..., -0.11076976,
        -0.22573306, -0.15762208],
       ...,
       [-0.96968969, -0.99366033, -0.82701582, ..., -0.11076976,
        -0.22573306, -0.15762208],
       [ 1.05853018,  2.61464637,  0.84372321, ..., -0.11076976,
        -0.22573306, -0.15762208],
       [ 1.97330695, -0.05645153,  2.38273089, ..., -0.11076976,
        -0.22573306, -0.15762208]])

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [67]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
model

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [68]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred=model.predict(X_test)
y_pred

array([ 58.23545484,  90.35947828,  78.58115797,  76.3536189 ,
        70.03135328,  76.03916578,  45.07627515,  46.84812574,
        93.32920484,  83.19688062,  51.12119703,  79.10020094,
        51.10508375,  71.44932203,  62.86436109,  92.86875562,
        37.13193922,  72.58201246,  61.3692439 ,  41.29063062,
       113.7286189 ,  36.76914625,  53.3043025 ,  49.25400953,
        48.48814039,  62.79392652,  77.31260328,  82.42881422,
        64.15513258,  66.437115  ,  35.28477125,  61.41416578,
        79.8692439 ,  56.94785719,  75.24326734,  49.24766187,
        23.21201734,  29.13682203,  61.7261775 ,  37.54404859,
        47.90342359,  61.08750562,  85.78623609,  72.01865797,
        72.0978816 ,  57.5029109 ,  64.33970289, 164.15147047,
        38.14414625,  62.90488844,  68.47910719,  59.30576734,
        53.19138746,  45.70713453,  58.96397047,  42.22178297,
        87.33311109,  38.57188062, 118.59482984, 118.79209547,
        56.14963941,  56.06272535,  42.10899   ,  79.74

## Review Model Outcomes — Iterate over additional models as needed

In [69]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from . 
# You can plug y_test and y_pred into the functions to evaluate the model

In [70]:
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error
explained_variance_score(y_test, y_pred)


0.9277955729035006

In [71]:
mean_absolute_error(y_test, y_pred)

5.210503138167747

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [72]:
model.intercept_

64.13193921635681

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [73]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


Coefficient
summit_elev        2.087404e+01
vertical_drop      6.484792e+00
base_elev          1.911799e+01
trams              9.072607e+11
fastEight          8.919262e+10
fastSixes          1.055902e+12
fastQuads          3.561817e+12
quad               2.126183e+12
triple             2.623418e+12
double             2.940825e+12
surface            3.337155e+12
total_chairs       9.395399e+12
Runs               1.292419e+00
TerrainParks       2.365646e-01
LongestRun_mi      4.038086e-01
SkiableTerrain_ac  1.855469e-01
Snow Making_ac     9.132690e-01
daysOpenLastYear   2.359619e-01
yearsOpen          3.853455e-01
averageSnowfall    2.460938e-01
AdultWeekday       1.914087e+01
projectedDaysOpen  8.361816e-02
NightSkiing_ac     7.617188e-01
clusters           1.087158e+00
Arizona            3.084717e-01
California         4.660034e-01
Colorado           7.250977e-01
Connecticut        8.902588e-01
Idaho              5.158691e-01
Illinois           3.833008e-01
Indiana            2.060852e-01
Iowa               1.173706e-01
Maine              9.013672e-01
Maryland           7.286377e-01
Massachusetts      2.035767e+00
Michigan           7.115479e-01
Minnesota          5.053711e-02
Missouri           5.395508e-02
Montana            1.071236e+00
Nevada             2.934570e-01
New Hampshire      2.107452e+00
New Jersey         3.164062e-01
New Mexico         1.216797e+00
New York           9.818115e-01
North Carolina     2.239868e+00
Ohio               4.112854e-01
Oregon             3.270874e-01
Pennsylvania       1.408417e+00
Rhode Island       3.715820e-01
South Dakota       1.820679e-01
Tennessee          1.209900e+00
Utah               2.591553e-01
Vermont            1.064270e+00
Virginia           1.086548e+00
Washington         9.619141e-02
West Virginia      1.532364e+00
Wisconsin          8.242188e-01
Wyoming            4.898682e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [74]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df

X2=X.iloc[:,:-35]
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X2_scaled=scaler.transform(X2) 

In [75]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X2_train, X2_test, y_train, y_test = train_test_split(X2_scaled, y, test_size=0.25, random_state=1)
X2_train

array([[-1.58669848e-01, -9.14414942e-01,  4.04832918e-02, ...,
        -7.92746293e-01, -1.04716512e+00, -6.09611897e-01],
       [-2.55187912e-01,  1.47351277e+00, -7.53117745e-01, ...,
         8.83519743e-01,  5.20812283e-01, -6.09611897e-01],
       [-9.76392337e-01, -9.67245201e-01, -8.75210213e-01, ...,
         0.00000000e+00, -9.90324405e-16, -3.31010103e-01],
       ...,
       [-9.69689694e-01, -9.93660331e-01, -8.27015818e-01, ...,
        -7.50839642e-01, -6.98725699e-01, -2.34571021e-01],
       [ 1.05853018e+00,  2.61464637e+00,  8.43723208e-01, ...,
         0.00000000e+00,  6.44428248e+00,  2.60502419e+00],
       [ 1.97330695e+00, -5.64515333e-02,  2.38273089e+00, ...,
         1.30258625e+00,  5.55656226e-01, -6.09611897e-01]])

In [76]:
lm = linear_model.LinearRegression()
model = lm.fit(X2_train,y_train)
model

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [77]:
y_pred=model.predict(X2_test)
y_pred

array([ 57.82068231,  84.35091684,  81.87613152,  78.37466721,
        67.20068276,  80.35013956,  37.64594234,  47.4278293 ,
        88.62784799,  85.73167821,  51.85879712,  79.30763066,
        53.87597191,  65.43146533,  63.29671914,  89.08855885,
        37.64533103,  76.70814467,  64.7447591 ,  43.33053233,
       108.12426146,  37.08607955,  53.47410038,  49.89835267,
        50.52149628,  63.75952496,  66.33530511,  75.50586238,
        64.10632783,  64.55902618,  39.54269806,  52.2490845 ,
        78.03420667,  56.96457192,  75.22327002,  54.16449448,
        25.19692674,  27.95475659,  55.11318002,  38.91498993,
        49.750731  ,  64.55405899,  88.55560489,  66.80180621,
        68.70688333,  53.65200941,  62.41758541, 168.26056813,
        30.47554359,  63.41269849,  62.97157613,  60.07556108,
        53.52364665,  48.54967873,  61.22425653,  42.98349852,
        86.89167923,  31.17395099, 121.83200434, 115.04510448,
        57.6119309 ,  54.21861246,  39.45565952,  86.65

In [78]:
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error
explained_variance_score(y_test, y_pred)

0.9220831169611157

In [79]:
mean_absolute_error(y_test, y_pred)

5.56568262579038

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

In [80]:
model.intercept_

64.06004480314571

In [81]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X2.columns, columns=['Coefficient'])



Coefficient
summit_elev          13.034342
vertical_drop         4.459066
base_elev            10.139237
trams                 0.233177
fastEight             0.050862
fastSixes             0.322328
fastQuads             0.687531
quad                  1.516598
triple                1.360779
double                0.313228
surface               1.233862
total_chairs          0.656825
Runs                  1.116846
TerrainParks          0.242047
LongestRun_mi         0.317966
SkiableTerrain_ac     0.244761
Snow Making_ac        0.613794
daysOpenLastYear      1.091645
yearsOpen             0.282277
averageSnowfall       1.833735
AdultWeekday         20.157048
projectedDaysOpen     0.426983
NightSkiing_ac        0.343832

#### Model 3

In [82]:
X2.drop(['summit_elev','base_elev'],axis=1, inplace=True)
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 

y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X2_scaled=scaler.transform(X2) 


In [83]:
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X2_train, X2_test, y_train, y_test = train_test_split(X2_scaled, y, test_size=0.25, random_state=1)
X2_train

array([[-9.14414942e-01, -3.08939686e-01, -5.51317846e-02, ...,
        -7.92746293e-01, -1.04716512e+00, -6.09611897e-01],
       [ 1.47351277e+00, -3.08939686e-01, -5.51317846e-02, ...,
         8.83519743e-01,  5.20812283e-01, -6.09611897e-01],
       [-9.67245201e-01, -3.08939686e-01, -5.51317846e-02, ...,
         0.00000000e+00, -9.90324405e-16, -3.31010103e-01],
       ...,
       [-9.93660331e-01, -3.08939686e-01, -5.51317846e-02, ...,
        -7.50839642e-01, -6.98725699e-01, -2.34571021e-01],
       [ 2.61464637e+00, -3.08939686e-01, -5.51317846e-02, ...,
         0.00000000e+00,  6.44428248e+00,  2.60502419e+00],
       [-5.64515333e-02, -3.08939686e-01, -5.51317846e-02, ...,
         1.30258625e+00,  5.55656226e-01, -6.09611897e-01]])

In [84]:
lm = linear_model.LinearRegression()
model = lm.fit(X2_train,y_train)
model

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [85]:
y_pred=model.predict(X2_test)
y_pred

array([ 58.22815714,  84.06391731,  83.03560778,  78.533348  ,
        66.09354247,  81.37061506,  37.39538654,  47.10292907,
        88.36234513,  85.22668043,  51.71217691,  80.57652367,
        54.05502054,  65.11012705,  63.04379078,  88.67388821,
        40.5665392 ,  77.57017078,  66.42602133,  43.15221604,
       107.64594472,  36.74256107,  52.59499341,  49.78231297,
        50.39060642,  63.80766161,  66.56893371,  72.5305692 ,
        64.75134894,  64.43549111,  40.52573771,  52.37773682,
        78.34289172,  56.4034838 ,  74.7756329 ,  54.13293094,
        25.04297003,  27.6779828 ,  54.93790824,  38.54356054,
        50.21242657,  64.89477437,  88.37773784,  66.40773987,
        68.9350966 ,  53.04290972,  63.05371455, 168.28072478,
        30.40035133,  64.5064958 ,  62.58619586,  60.94038096,
        53.5301541 ,  48.3788436 ,  61.32513305,  43.3982614 ,
        86.09087986,  31.81597431, 120.56855287, 114.78838025,
        57.24933799,  54.79383277,  39.14409052,  86.14

In [86]:
explained_variance_score(y_test, y_pred)

0.9243057470047055

In [87]:
mean_absolute_error(y_test, y_pred)

5.538078961917421

In [88]:
model.intercept_

64.06166833087156

In [89]:
pd.DataFrame(abs(lm.coef_), X2.columns, columns=['Coefficient'])

Coefficient
vertical_drop         1.110726
trams                 0.212294
fastEight             0.103876
fastSixes             0.327224
fastQuads             0.770725
quad                  1.554804
triple                1.330670
double                0.320582
surface               1.215124
total_chairs          0.701596
Runs                  1.307364
TerrainParks          0.157353
LongestRun_mi         0.296431
SkiableTerrain_ac     0.331384
Snow Making_ac        0.504454
daysOpenLastYear      0.967384
yearsOpen             0.302607
averageSnowfall       2.180381
AdultWeekday         20.082559
projectedDaysOpen     0.195712
NightSkiing_ac        0.480682

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.9277955729035006 | 5.210503138167747 |-|
| Model 2. | 0.0.9220831169611157| 5.56568262579038 |'state'|
| Model 3. | 0.9243057470047055 | 5.538078961917421 |'state','summit_elev','base_elev'|


Model Selection:

In [90]:
#Model 1